simulation/test_utp.cpp

```diff
@@ -139,7 +139,7 @@ TORRENT_TEST(utp_pmtud)
 	TEST_EQUAL(metric(cnt, "utp.utp_packets_in"), 593);
 	TEST_EQUAL(metric(cnt, "utp.utp_payload_pkts_in"), 66);
 
-	TEST_EQUAL(metric(cnt, "utp.utp_packets_out"), 603);
+	TEST_EQUAL(metric(cnt, "utp.utp_packets_out"), 602);
```

Comment:
it's not clear to me why this change would result in one less packet to be sent. If anything, a larger SACK header would eat into the size of the payload that would fit in the packet

Reply:
I'd only updated that packet count in order to see if the other tests would still pass, which they seem to do.

I'm yet to work out why that particular test has one less packet, but that was my first thought as well - with the extra 3 bytes possible in the header, if anything it should be the other way round.

Reply:
There might be an issue when picking up a Nagled packet with a SACK already attached if the sack size has increased since it was held back. It appears to keep the old size, rather than growing it if needed. It looks like an existing issue that just happened to trigger on that one test with this PR compared to without.

In this case, the unpatched version misses out on acking a packet (the 1 byte SACK payload needs to grow to hold 9 bytes in this case but doesn't), while the patched version already has enough padding to grow (although it will be an issue growing a Nagled SACK from 32-33 bytes). As a result, the unpatched version sends an extra ack packet later on to catch up.

When comparing utp debug logfiles of the failing test using this PR against those without they appear to diverge right after that happens.

Hopefully if that gets fixed, then that test will pass without needing to update the packet count.

Reply:
Can you attach the two pcap one with 603 packets, one with 602?